// @okxweb3/coin-aptos/dist/v2/transactions/instances/moduleId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModuleId = void 0;
const serializer_1 = require("../../bcs/serializer");
const core_1 = require("../../core");
const identifier_1 = require("./identifier");
class ModuleId extends serializer_1.Serializable {
    constructor(address, name) {
        super();
        this.address = address;
        this.name = name;
    }
    static fromStr(moduleId) {
        const parts = moduleId.split("::");
        if (parts.length !== 2) {
            throw new Error("Invalid module id.");
        }
        return new ModuleId(core_1.AccountAddress.fromString(parts[0]), new identifier_1.Identifier(parts[1]));
    }
    serialize(serializer) {
        this.address.serialize(serializer);
        this.name.serialize(serializer);
    }
    static deserialize(deserializer) {
        const address = core_1.AccountAddress.deserialize(deserializer);
        const name = identifier_1.Identifier.deserialize(deserializer);
        return new ModuleId(address, name);
    }
}
exports.ModuleId = ModuleId;
