// @okxweb3/coin-aptos/dist/v2/transactions/instances/multiAgentTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultiAgentTransaction = void 0;
const serializer_1 = require("../../bcs/serializer");
const core_1 = require("../../core");
const rawTransaction_1 = require("./rawTransaction");
class MultiAgentTransaction extends serializer_1.Serializable {
    constructor(rawTransaction, secondarySignerAddresses, feePayerAddress) {
        super();
        this.rawTransaction = rawTransaction;
        this.feePayerAddress = feePayerAddress;
        this.secondarySignerAddresses = secondarySignerAddresses;
    }
    serialize(serializer) {
        this.rawTransaction.serialize(serializer);
        serializer.serializeVector(this.secondarySignerAddresses);
        if (this.feePayerAddress === undefined) {
            serializer.serializeBool(false);
        }
        else {
            serializer.serializeBool(true);
            this.feePayerAddress.serialize(serializer);
        }
    }
    static deserialize(deserializer) {
        const rawTransaction = rawTransaction_1.RawTransaction.deserialize(deserializer);
        const secondarySignerAddresses = deserializer.deserializeVector(core_1.AccountAddress);
        const feepayerPresent = deserializer.deserializeBool();
        let feePayerAddress;
        if (feepayerPresent) {
            feePayerAddress = core_1.AccountAddress.deserialize(deserializer);
        }
        return new MultiAgentTransaction(rawTransaction, secondarySignerAddresses, feePayerAddress);
    }
}
exports.MultiAgentTransaction = MultiAgentTransaction;
