// @okxweb3/coin-aptos/dist/v2/transactions/instances/rotationProofChallenge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RotationProofChallenge = void 0;
const serializer_1 = require("../../bcs/serializer");
const accountAddress_1 = require("../../core/accountAddress");
const bcs_1 = require("../../bcs");
class RotationProofChallenge extends serializer_1.Serializable {
    constructor(args) {
        super();
        this.accountAddress = accountAddress_1.AccountAddress.ONE;
        this.moduleName = new bcs_1.MoveString("account");
        this.structName = new bcs_1.MoveString("RotationProofChallenge");
        this.sequenceNumber = new bcs_1.U64(args.sequenceNumber);
        this.originator = args.originator;
        this.currentAuthKey = args.currentAuthKey;
        this.newPublicKey = bcs_1.MoveVector.U8(args.newPublicKey.toUint8Array());
    }
    serialize(serializer) {
        serializer.serialize(this.accountAddress);
        serializer.serialize(this.moduleName);
        serializer.serialize(this.structName);
        serializer.serialize(this.sequenceNumber);
        serializer.serialize(this.originator);
        serializer.serialize(this.currentAuthKey);
        serializer.serialize(this.newPublicKey);
    }
}
exports.RotationProofChallenge = RotationProofChallenge;
