// @okxweb3/coin-aptos/dist/v2/internal/coin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transferCoinTransaction = void 0;
const const_1 = require("../utils/const");
const transactionSubmission_1 = require("./transactionSubmission");
const transactions_1 = require("../transactions");
const coinTransferAbi = {
    typeParameters: [{ constraints: [] }],
    parameters: [new transactions_1.TypeTagAddress(), new transactions_1.TypeTagU64()],
};
async function transferCoinTransaction(args) {
    const { aptosConfig, sender, recipient, amount, coinType, options } = args;
    const coinStructType = coinType ?? const_1.APTOS_COIN;
    return (0, transactionSubmission_1.generateTransaction)({
        aptosConfig,
        sender,
        data: {
            function: "0x1::aptos_account::transfer_coins",
            typeArguments: [coinStructType],
            functionArguments: [recipient, amount],
            abi: coinTransferAbi,
        },
        options,
    });
}
exports.transferCoinTransaction = transferCoinTransaction;
