// @okxweb3/coin-aptos/dist/v2/internal/fungibleAsset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transferFungibleAsset = void 0;
const transactions_1 = require("../transactions");
const transactionSubmission_1 = require("./transactionSubmission");
const faTransferAbi = {
    typeParameters: [{ constraints: [] }],
    parameters: [(0, transactions_1.parseTypeTag)("0x1::object::Object"), new transactions_1.TypeTagAddress(), new transactions_1.TypeTagU64()],
};
async function transferFungibleAsset(args) {
    const { aptosConfig, sender, fungibleAssetMetadataAddress, recipient, amount, options } = args;
    return (0, transactionSubmission_1.generateTransaction)({
        aptosConfig,
        sender: sender.accountAddress,
        data: {
            function: "0x1::primary_fungible_store::transfer",
            typeArguments: ["0x1::fungible_asset::Metadata"],
            functionArguments: [fungibleAssetMetadataAddress, recipient, amount],
            abi: faTransferAbi,
        },
        options,
    });
}
exports.transferFungibleAsset = transferFungibleAsset;
