// @okxweb3/coin-aptos/dist/aptos_account.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAddressFromAccountOrAddress = exports.AptosAccount = void 0;
const hex_string_1 = require("./hex_string");
const tweetnacl_1 = __importDefault(require("tweetnacl"));
const bip39 = __importStar(require("@scure/bip39"));
const utils_1 = require("@noble/hashes/utils");
const sha256_1 = require("@noble/hashes/sha256");
const sha3_1 = require("@noble/hashes/sha3");
const hd_key_1 = require("./utils/hd_key");
const utils_2 = require("./utils");
const aptos_types_1 = require("./transaction_builder/aptos_types");
const bcs_1 = require("./transaction_builder/bcs");
const util_1 = require("util");
class AptosAccount {
    static fromAptosAccountObject(obj) {
        return new AptosAccount(hex_string_1.HexString.ensure(obj.privateKeyHex).toUint8Array(), obj.address);
    }
    static fromPrivateKey(privateKey) {
        return new AptosAccount(privateKey.toUint8Array());
    }
    static isValidPath(path) {
        return /^m\/44'\/637'\/[0-9]+'\/[0-9]+'\/[0-9]+'+$/.test(path);
    }
    static fromDerivePath(path, mnemonics) {
        if (!AptosAccount.isValidPath(path)) {
            throw new Error("Invalid derivation path");
        }
        const normalizeMnemonics = mnemonics
            .trim()
            .split(/\s+/)
            .map((part) => part.toLowerCase())
            .join(" ");
        const { key } = (0, hd_key_1.derivePath)(path, (0, utils_1.bytesToHex)(bip39.mnemonicToSeedSync(normalizeMnemonics)));
        return new AptosAccount(key);
    }
    constructor(privateKeyBytes, address) {
        if (!privateKeyBytes || (privateKeyBytes.length !== 32 && privateKeyBytes.length !== 64)) {
            throw new Error('invalid key');
        }
        if (privateKeyBytes) {
            this.signingKey = tweetnacl_1.default.sign.keyPair.fromSeed(privateKeyBytes.slice(0, 32));
        }
        else {
            this.signingKey = tweetnacl_1.default.sign.keyPair();
        }
        this.accountAddress = hex_string_1.HexString.ensure(address || this.authKey().hex());
    }
    address() {
        return this.accountAddress;
    }
    authKey() {
        const pubKey = new aptos_types_1.Ed25519PublicKey(this.signingKey.publicKey);
        const authKey = aptos_types_1.AuthenticationKey.fromEd25519PublicKey(pubKey);
        return authKey.derivedAddress();
    }
    static getResourceAccountAddress(sourceAddress, seed) {
        const source = (0, bcs_1.bcsToBytes)(aptos_types_1.AccountAddress.fromHex(sourceAddress));
        const bytes = new Uint8Array([...source, ...seed, aptos_types_1.AuthenticationKey.DERIVE_RESOURCE_ACCOUNT_SCHEME]);
        const hash = sha3_1.sha3_256.create();
        hash.update(bytes);
        return hex_string_1.HexString.fromUint8Array(hash.digest());
    }
    static getCollectionID(creatorAddress, collectionName) {
        const seed = new util_1.TextEncoder().encode(`${creatorAddress}::${collectionName}`);
        const hash = sha256_1.sha256.create();
        hash.update(seed);
        return hex_string_1.HexString.fromUint8Array(hash.digest());
    }
    pubKey() {
        return hex_string_1.HexString.fromUint8Array(this.signingKey.publicKey);
    }
    signBuffer(buffer) {
        const signature = tweetnacl_1.default.sign.detached(buffer, this.signingKey.secretKey);
        return hex_string_1.HexString.fromUint8Array(signature);
    }
    signHexString(hexString) {
        const toSign = hex_string_1.HexString.ensure(hexString).toUint8Array();
        return this.signBuffer(toSign);
    }
    verifySignature(message, signature) {
        const rawMessage = hex_string_1.HexString.ensure(message).toUint8Array();
        const rawSignature = hex_string_1.HexString.ensure(signature).toUint8Array();
        return tweetnacl_1.default.sign.detached.verify(rawMessage, rawSignature, this.signingKey.publicKey);
    }
    toPrivateKeyObject() {
        return {
            address: this.address().hex(),
            publicKeyHex: this.pubKey().hex(),
            privateKeyHex: hex_string_1.HexString.fromUint8Array(this.signingKey.secretKey.slice(0, 32)).hex(),
        };
    }
}
__decorate([
    (0, utils_2.Memoize)()
], AptosAccount.prototype, "authKey", null);
exports.AptosAccount = AptosAccount;
function getAddressFromAccountOrAddress(accountOrAddress) {
    return accountOrAddress instanceof AptosAccount ? accountOrAddress.address() : hex_string_1.HexString.ensure(accountOrAddress);
}
exports.getAddressFromAccountOrAddress = getAddressFromAccountOrAddress;
