// @okxweb3/coin-aptos/dist/hex_string.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HexString = void 0;
const utils_1 = require("@noble/hashes/utils");
class HexString {
    static fromBuffer(buffer) {
        return HexString.fromUint8Array(buffer);
    }
    static fromUint8Array(arr) {
        return new HexString((0, utils_1.bytesToHex)(arr));
    }
    static ensure(hexString) {
        if (typeof hexString === "string") {
            return new HexString(hexString);
        }
        return hexString;
    }
    constructor(hexString) {
        if (hexString.startsWith("0x")) {
            this.hexString = hexString;
        }
        else {
            this.hexString = `0x${hexString}`;
        }
    }
    hex() {
        return this.hexString;
    }
    noPrefix() {
        return this.hexString.slice(2);
    }
    toString() {
        return this.hex();
    }
    toShortString() {
        const trimmed = this.hexString.replace(/^0x0*/, "");
        return `0x${trimmed}`;
    }
    toUint8Array() {
        return Uint8Array.from((0, utils_1.hexToBytes)(this.noPrefix()));
    }
}
exports.HexString = HexString;
