// @okxweb3/coin-aptos/dist/v2/utils/const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProcessorType = exports.RAW_TRANSACTION_WITH_DATA_SALT = exports.RAW_TRANSACTION_SALT = exports.APTOS_COIN = exports.DEFAULT_TXN_TIMEOUT_SEC = exports.DEFAULT_TXN_EXP_SEC_FROM_NOW = exports.DEFAULT_MAX_GAS_AMOUNT = exports.AptosApiType = void 0;
var AptosApiType;
(function (AptosApiType) {
    AptosApiType["FULLNODE"] = "Fullnode";
    AptosApiType["INDEXER"] = "Indexer";
    AptosApiType["FAUCET"] = "Faucet";
})(AptosApiType = exports.AptosApiType || (exports.AptosApiType = {}));
exports.DEFAULT_MAX_GAS_AMOUNT = 200000;
exports.DEFAULT_TXN_EXP_SEC_FROM_NOW = 20;
exports.DEFAULT_TXN_TIMEOUT_SEC = 20;
exports.APTOS_COIN = "0x1::aptos_coin::AptosCoin";
exports.RAW_TRANSACTION_SALT = "APTOS::RawTransaction";
exports.RAW_TRANSACTION_WITH_DATA_SALT = "APTOS::RawTransactionWithData";
var ProcessorType;
(function (ProcessorType) {
    ProcessorType["ACCOUNT_TRANSACTION_PROCESSOR"] = "account_transactions_processor";
    ProcessorType["DEFAULT"] = "default_processor";
    ProcessorType["EVENTS_PROCESSOR"] = "events_processor";
    ProcessorType["FUNGIBLE_ASSET_PROCESSOR"] = "fungible_asset_processor";
    ProcessorType["STAKE_PROCESSOR"] = "stake_processor";
    ProcessorType["TOKEN_V2_PROCESSOR"] = "token_v2_processor";
    ProcessorType["USER_TRANSACTION_PROCESSOR"] = "user_transaction_processor";
})(ProcessorType = exports.ProcessorType || (exports.ProcessorType = {}));
