// @okxweb3/coin-starknet/dist/utils/json/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUnsafeNumberReason = exports.toSafeNumberOrThrow = exports.isSafeNumber = exports.isNumber = exports.isInteger = exports.UnsafeNumberReason = exports.parseNumberAndBigInt = exports.parseLosslessNumber = exports.toLosslessNumber = exports.isLosslessNumber = exports.LosslessNumber = exports.stringify = exports.parse = void 0;
var parse_1 = require("./parse");
Object.defineProperty(exports, "parse", { enumerable: true, get: function () { return parse_1.parse; } });
var stringify_1 = require("./stringify");
Object.defineProperty(exports, "stringify", { enumerable: true, get: function () { return stringify_1.stringify; } });
var LosslessNumber_1 = require("./LosslessNumber");
Object.defineProperty(exports, "LosslessNumber", { enumerable: true, get: function () { return LosslessNumber_1.LosslessNumber; } });
Object.defineProperty(exports, "isLosslessNumber", { enumerable: true, get: function () { return LosslessNumber_1.isLosslessNumber; } });
Object.defineProperty(exports, "toLosslessNumber", { enumerable: true, get: function () { return LosslessNumber_1.toLosslessNumber; } });
var numberParsers_1 = require("./numberParsers");
Object.defineProperty(exports, "parseLosslessNumber", { enumerable: true, get: function () { return numberParsers_1.parseLosslessNumber; } });
Object.defineProperty(exports, "parseNumberAndBigInt", { enumerable: true, get: function () { return numberParsers_1.parseNumberAndBigInt; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "UnsafeNumberReason", { enumerable: true, get: function () { return utils_1.UnsafeNumberReason; } });
Object.defineProperty(exports, "isInteger", { enumerable: true, get: function () { return utils_1.isInteger; } });
Object.defineProperty(exports, "isNumber", { enumerable: true, get: function () { return utils_1.isNumber; } });
Object.defineProperty(exports, "isSafeNumber", { enumerable: true, get: function () { return utils_1.isSafeNumber; } });
Object.defineProperty(exports, "toSafeNumberOrThrow", { enumerable: true, get: function () { return utils_1.toSafeNumberOrThrow; } });
Object.defineProperty(exports, "getUnsafeNumberReason", { enumerable: true, get: function () { return utils_1.getUnsafeNumberReason; } });
__exportStar(require("./types"), exports);
