// @okxweb3/coin-starknet/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.accountClassHash = exports.ProxyAccountClassHash = exports.ETHBridge = exports.ETH = exports.UDC = exports.TransactionHashPrefix = exports.StarknetChainId = exports.NetworkName = exports.API_VERSION = exports.MASK_251 = exports.MASK_250 = exports.ZERO = void 0;
exports.ZERO = 0n;
exports.MASK_250 = 2n ** 250n - 1n;
exports.MASK_251 = 2n ** 251n;
exports.API_VERSION = exports.ZERO;
var NetworkName;
(function (NetworkName) {
    NetworkName["SN_MAIN"] = "SN_MAIN";
    NetworkName["SN_GOERLI"] = "SN_GOERLI";
    NetworkName["SN_GOERLI2"] = "SN_GOERLI2";
})(NetworkName = exports.NetworkName || (exports.NetworkName = {}));
var StarknetChainId;
(function (StarknetChainId) {
    StarknetChainId["SN_MAIN"] = "0x534e5f4d41494e";
    StarknetChainId["SN_GOERLI"] = "0x534e5f474f45524c49";
    StarknetChainId["SN_GOERLI2"] = "0x534e5f474f45524c4932";
})(StarknetChainId = exports.StarknetChainId || (exports.StarknetChainId = {}));
var TransactionHashPrefix;
(function (TransactionHashPrefix) {
    TransactionHashPrefix["DECLARE"] = "0x6465636c617265";
    TransactionHashPrefix["DEPLOY"] = "0x6465706c6f79";
    TransactionHashPrefix["DEPLOY_ACCOUNT"] = "0x6465706c6f795f6163636f756e74";
    TransactionHashPrefix["INVOKE"] = "0x696e766f6b65";
    TransactionHashPrefix["L1_HANDLER"] = "0x6c315f68616e646c6572";
})(TransactionHashPrefix = exports.TransactionHashPrefix || (exports.TransactionHashPrefix = {}));
exports.UDC = {
    ADDRESS: '0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf',
    ENTRYPOINT: 'deployContract',
};
exports.ETH = "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7";
exports.ETHBridge = "0x073314940630fd6dcda0d772d4c972c4e0a9946bef9dabf4ef84eda8ef542b82";
exports.ProxyAccountClassHash = "0x3530cc4759d78042f1b543bf797f5f3d647cde0388c33734cf91b7f7b9314a9";
exports.accountClassHash = "0x309c042d3729173c7f2f91a34f04d8c509c1b292d334679ef1aabf8da0899cc";
