// @okxweb3/coin-aptos/dist/transaction_builder/aptos_types/account_address.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountAddress = void 0;
const hex_string_1 = require("../../hex_string");
class AccountAddress {
    constructor(address) {
        if (address.length !== AccountAddress.LENGTH) {
            throw new Error("Expected address of length 32");
        }
        this.address = address;
    }
    static fromHex(addr) {
        let address = hex_string_1.HexString.ensure(addr);
        if (address.noPrefix().length % 2 !== 0) {
            address = new hex_string_1.HexString(`0${address.noPrefix()}`);
        }
        const addressBytes = address.toUint8Array();
        if (addressBytes.length > AccountAddress.LENGTH) {
            throw new Error("Hex string is too long. Address's length is 32 bytes.");
        }
        else if (addressBytes.length === AccountAddress.LENGTH) {
            return new AccountAddress(addressBytes);
        }
        const res = new Uint8Array(AccountAddress.LENGTH);
        res.set(addressBytes, AccountAddress.LENGTH - addressBytes.length);
        return new AccountAddress(res);
    }
    static isValid(addr) {
        if (addr === "") {
            return false;
        }
        let address = hex_string_1.HexString.ensure(addr);
        if (address.noPrefix().length % 2 !== 0) {
            address = new hex_string_1.HexString(`0${address.noPrefix()}`);
        }
        const addressBytes = address.toUint8Array();
        return addressBytes.length <= AccountAddress.LENGTH;
    }
    toHexString() {
        return hex_string_1.HexString.fromUint8Array(this.address).hex();
    }
    serialize(serializer) {
        serializer.serializeFixedBytes(this.address);
    }
    static deserialize(deserializer) {
        return new AccountAddress(deserializer.deserializeFixedBytes(AccountAddress.LENGTH));
    }
    static standardizeAddress(address) {
        const lowercaseAddress = address.toLowerCase();
        const addressWithoutPrefix = lowercaseAddress.startsWith("0x") ? lowercaseAddress.slice(2) : lowercaseAddress;
        const addressWithPadding = addressWithoutPrefix.padStart(64, "0");
        return `0x${addressWithPadding}`;
    }
}
exports.AccountAddress = AccountAddress;
AccountAddress.LENGTH = 32;
AccountAddress.CORE_CODE_ADDRESS = AccountAddress.fromHex("0x1");
