// @okxweb3/coin-aptos/dist/v2/api/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const transactionSubmission_1 = require("../internal/transactionSubmission");
const build_1 = require("./transactionSubmission/build");
const simulate_1 = require("./transactionSubmission/simulate");
const submit_1 = require("./transactionSubmission/submit");
class Transaction {
    constructor(config) {
        this.config = config;
        this.build = new build_1.Build(this.config);
        this.simulate = new simulate_1.Simulate(this.config);
        this.submit = new submit_1.Submit(this.config);
    }
    getSigningMessage(args) {
        return (0, transactionSubmission_1.getSigningMessage)(args);
    }
    async publishPackageTransaction(args) {
        return (0, transactionSubmission_1.publicPackageTransaction)({ aptosConfig: this.config, ...args });
    }
    async rotateAuthKey(args) {
        return (0, transactionSubmission_1.rotateAuthKey)({ aptosConfig: this.config, ...args });
    }
    sign(args) {
        return (0, transactionSubmission_1.signTransaction)({
            ...args,
        });
    }
    signAsFeePayer(args) {
        const { signer, transaction } = args;
        if (!transaction.feePayerAddress) {
            throw new Error(`Transaction ${transaction} is not a Fee Payer transaction`);
        }
        transaction.feePayerAddress = signer.accountAddress;
        return (0, transactionSubmission_1.signTransaction)({
            signer,
            transaction,
        });
    }
    async signAndSubmitTransaction(args) {
        const { signer, transaction } = args;
        return (0, transactionSubmission_1.signAndSubmitTransaction)({
            aptosConfig: this.config,
            signer,
            transaction,
        });
    }
}
exports.Transaction = Transaction;
